src/lib/utils/int_utils.h

```diff
@@ -0,0 +1,57 @@
+/*
+* (C) 2024 Jack Lloyd
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#ifndef BOTAN_INT_UTILS_H_
+#define BOTAN_INT_UTILS_H_
+
+#include <botan/concepts.h>
+#include <botan/types.h>
+#include <optional>
+
+namespace Botan {
+
+template <std::unsigned_integral T>
+constexpr inline std::optional<T> checked_add(T a, T b) {
+   const T r = a + b;
+   if(r - a != b) {
+      return {};
+   }
+   return r;
+}
+
+template <std::unsigned_integral T, std::unsigned_integral... Ts>
+   requires all_same_v<T, Ts...>
+constexpr inline std::optional<T> checked_add(T a, T b, Ts... rest) {
+   if(auto r = checked_add(a, b)) {
+      return checked_add(r.value(), rest...);
+   } else {
+      return {};
+   }
+}
+
+template <std::unsigned_integral T>
+constexpr inline std::optional<T> checked_mul(T a, T b) {
+   const T r = a * b;
+   // If a == 0 then the multiply certainly did not overflow
+   // Otherwise r / a == b unless overflow occured
+   if(a != 0 && r / a != b) {
+      return {};
+   }
+   return r;
+}
+
+template <std::unsigned_integral T, std::unsigned_integral RT>
+std::optional<RT> checked_cast(T input) {
+   const RT cast = static_cast<RT>(input);
+   if(static_cast<T>(cast) != input) {
+      return {};
+   }
+   return cast;
+}
```

Comment:
When the user is supposed to partially define the template at the call-site, the to-be-defined template parameters should come first, no?

If we wanted to overengineer this a little: C++20 seems to allow mixing fully-fledged template parameters and the new `auto` parameter syntax. That way, the user is forced to rely on type deduction for the function parameter, and also there's no confusion in the order of the template parameters:

```suggestion
template <std::unsigned_integral ReturnT>
std::optional<ReturnT> checked_cast(std::unsigned_integral auto input) {
   if(std::numeric_limits<ReturnT>::max() < input) {
      return {};
   }
   return static_cast<ReturnT>(input);
}
```

(Note that this might require `#include <limits>`. Or did you have something specific in mind when implementing it using the reversed cast for the check?)

Reply:
> When the user is supposed to partially define the template at the call-site, the to-be-defined template parameters should come first, no?

Yes - I did this as a cleanup of something that existed but was unused in `safeint.h` and then didn't think about the template parameter ordering.

> Or did you have something specific in mind when implementing it using the reversed cast for the check?

No, that's just how the preexisting code did it and it seemed reasonable. Your approach seems to lead to an interface that's easier to use correctly.